### PR TITLE
Fix SSO on applications with shared uid

### PR DESCRIPTION
**Pull request by erfanoabdi**
On original behavior of SSO first Nextcloud main app fetches the sso client package name by using `getCallingActivity` which is accurate and returns back the correct value and proceeds to token generation and so on: https://github.com/nextcloud/android/blob/6103234a21e0b89934a11886aa4b7ecb2b2c6a7e/app/src/main/java/com/owncloud/android/ui/activity/SsoGrantPermissionActivity.java#L93

But afterwards in token validation app is using `Binder.getCallingUid` which is the only way to obtain the caller package name, but in this case some SSO client applications that are using shared uid `android:sharedUserId` fails because `getNameForUid` returns only the first matched results in installed packages

So basically this PR just iterates through all the apps with matched UID and validates the token

More info: https://stackoverflow.com/questions/12918731/how-to-get-application-package-name-or-uid-which-is-trying-to-bind-my-service-fr

_Fix for the following error:_
```
com.nextcloud.android.sso.exceptions.TokenMismatchException: Provided authentication token does not match or package name is not permitted to perform this action
```

- [x] Tests written, or not not needed

### CODE REVIEW

**Review by alperozturk96 — APPROVED** [review]

**Review by ZetaTom — COMMENTED** [review]
Thank you for taking the time to fix this. In my testing, this change has worked well.

Please note that we have a check in place to block any commits that don't include a Signed-off-by line in the commit message. Refer to the [appropriate check](https://github.com/nextcloud/android/pull/12179/checks?check_run_id=18743499110) for more information, and a guide on how to sign off commits for this pull request.

**Comment by erfanoabdi** [comment]
Thanks @ZetaTom , added the sign off line

**Review by ZetaTom — APPROVED** [review]

**Review by tobiasKaminsky — APPROVED** [review]
